src/packages/contrib/contrib.cc

```diff
@@ -1610,116 +1610,131 @@ void f_replaceable(void) {
-        ignore[i + 2] = 0;
-      }
-    }
-    numignore += 2;
     obj = (sp - 1)->u.ob;
-  } else {
-    numignore = 2;
-    ignore = reinterpret_cast<char **>(DCALLOC(2, sizeof(char *), TAG_TEMPORARY, "replaceable"));
-    ignore[0] = findstring(APPLY_CREATE);
-    ignore[1] = findstring(APPLY___INIT);
+  }
+  else {
     obj = sp->u.ob;
   }
 
   prog = obj->prog;
-  num = prog->num_functions_defined + prog->last_inherited;
-
-  for (i = 0; i < num; i++) {
-    if (prog->function_flags[i] & (FUNC_INHERITED | FUNC_NO_CODE)) {
-      continue;
+  if (obj == simul_efun_ob || prog->func_ref) {
+    replaceable = 0;
+  }
+  else {
+    if (st_num_arg == 2) {
+      numignore = sp->u.arr->size;
+      if (numignore) {
+        ignore = reinterpret_cast<char **>(
+            DCALLOC(numignore + 2, sizeof(char *), TAG_TEMPORARY, "replaceable"));
+      } else {
+        ignore = 0;
+      }
+      ignore[0] = findstring(APPLY_CREATE);
+      ignore[1] = findstring(APPLY___INIT);
+      for (i = 0; i < numignore; i++) {
+        if (sp->u.arr->item[i].type == T_STRING) {
+          ignore[i + 2] = findstring(sp->u.arr->item[i].u.string);
+        } else {
+          ignore[i + 2] = 0;
+        }
+      }
+      numignore += 2;
+    } else {
+      numignore = 2;
+      ignore = reinterpret_cast<char **>(DCALLOC(2, sizeof(char *), TAG_TEMPORARY, "replaceable"));
+      ignore[0] = findstring(APPLY_CREATE);
+      ignore[1] = findstring(APPLY___INIT);
     }
-    for (j = 0; j < numignore; j++) {
-      if (ignore[j] == find_func_entry(prog, i)->funcname) {
+
+    num = prog->num_functions_defined + prog->last_inherited;
+
+    for (i = 0; i < num; i++) {
+      if (prog->function_flags[i] & (FUNC_INHERITED | FUNC_NO_CODE)) {
+        continue;
+      }
+      for (j = 0; j < numignore; j++) {
+        if (ignore[j] == find_func_entry(prog, i)->funcname) {
+          break;
+        }
+      }
+      if (j == numignore) {
         break;
       }
     }
-    if (j == numignore) {
-      break;
-    }
-  }
 
-  replaceable = (i == num);
-  if (obj == simul_efun_ob || prog->func_ref) {
-    replaceable = 0;
+    replaceable = (i == num);
+    FREE(ignore);
   }
 
   if (st_num_arg == 2) {
     free_array((sp--)->u.arr);
   }
-  FREE(ignore);
   free_svalue(sp, "f_replaceable");
   put_number(replaceable);
 }
 #endif
 
 #ifdef F_PROGRAM_INFO
+typedef struct {
+    int h_c,    // how many headers?
+        p_s,    // how big are the compiled programs?
+        ff_c,   // how many functions flags?
+        fd_c,   // how many function definitions?
+        s_c,    // how many strings?
+        v_c,    // how many variables?
+        i_c,    // how many inherits?
+        nc_c,   // how many classes (structs)?
+        mc_c,   // how many class members?
+        t_c,    // how mamy saved types (#pragma save_types)?
+        total;  // how big all together?
+} program_info_t;
+
+inline void sum_program_info(const program_t * const p, program_info_t &info) {
```

Comment:
_t can mean anything, it is just a legacy of the code, it doesn't justify
keep using C style in c++ codes.

Also in c++ typedef for struct is redundant, I've already removed all the
other cases, here should follow the code style too.

On Fri, Jan 8, 2016 at 12:09 AM, Dread Quixadhal <notifications@github.com>
wrote:

> In src/packages/contrib/contrib.cc
> <https://github.com/fluffos/fluffos/pull/297#discussion_r49089686>:
>
> >    free_svalue(sp, "f_replaceable");
> >    put_number(replaceable);
> >  }
> >  #endif
> >
> >  #ifdef F_PROGRAM_INFO
> > +typedef struct {
> > +    int h_c,    // how many headers?
> > +        p_s,    // how big are the compiled programs?
> > +        ff_c,   // how many functions flags?
> > +        fd_c,   // how many function definitions?
> > +        s_c,    // how many strings?
> > +        v_c,    // how many variables?
> > +        i_c,    // how many inherits?
> > +        nc_c,   // how many classes (structs)?
> > +        mc_c,   // how many class members?
> > +        t_c,    // how mamy saved types (#pragma save_types)?
> > +        total;  // how big all together?
> > +} program_info_t;
> > +
> > +inline void sum_program_info(const program_t * const p, program_info_t &info) {
>
> You guys do realize that the point of using a typedef is so you can have
> multiple uses all over the place without having to repeat the actual
> definition, right? IE: even if you *can* define your function with the raw
> struct definition right there, now if you make another function, you have
> to repeat the same definition again, and remember to change all occurances
> of it whenever it's updated, right? What's wrong with using typedef
> anyways? Don't forget to rename it everywhere if you ditch it... the _t at
> the end of the name is there to hint that it's a typedef, not a struct.
> … <#1521cd9e86ed7065_>
> On Thu, Jan 7, 2016 at 7:39 AM, Yucong Sun <notifications@github.com>
> wrote: In src/packages/contrib/contrib.cc <#297 (comment)
> <https://github.com/fluffos/fluffos/pull/297#discussion_r49068898>>: >
> free_svalue(sp, "f_replaceable"); > put_number(replaceable); > } > #endif >
> > #ifdef F_PROGRAM_INFO > +typedef struct { > + int h_c, // how many
> headers? > + p_s, // how big are the compiled programs? > + ff_c, // how
> many functions flags? > + fd_c, // how many function definitions? > + s_c,
> // how many strings? > + v_c, // how many variables? > + i_c, // how many
> inherits? > + nc_c, // how many classes (structs)? > + mc_c, // how many
> class members? > + t_c, // how mamy saved types (#pragma save_types)? > +
> total; // how big all together? > +} program_info_t; > + > +inline void
> sum_program_info(const program_t * const p, program_info_t &info) { yes,
> for recent C++, struct xxxx variable can be just replaced with xxx variable
> [image: image] <
> https://cloud.githubusercontent.com/assets/1256464/12170744/b8ed2e9c-b57e-11e5-89c6-6ff89b0d9787.png>
> — Reply to this email directly or view it on GitHub <
> https://github.com/fluffos/fluffos/pull/297/files#r49068898>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/fluffos/fluffos/pull/297/files#r49089686>.
>